### ISSUE TITLE
augmented assignments with larger indices involve unnecessary change of dtype

### ISSUE THREAD

**Issue by toobaz** [pydata/pandas#12755]
#### Code Sample, a copy-pastable example if possible

    In [2]: a = pd.Series(1, index=range(3))

    In [3]: b = pd.Series(2, index=range(4))

    In [4]: a.dtype
    Out[4]: dtype('int64')

    In [5]: a += b

    In [6]: a.dtype
    Out[6]: dtype('float64')

    In [7]: a.index == b.index
    ---------------------------------------------------------------------------
    ValueError                                Traceback (most recent call last)
    <ipython-input-8-2fae764c9047> in <module>()
    ----> 1 a.index == b.index

    /home/nobackup/repo/pandas/pandas/indexes/base.pyc in _evaluate_compare(self, other)
       3026                 if isinstance(other, (np.ndarray, Index, ABCSeries)):
       3027                     if other.ndim > 0 and len(self) != len(other):
    -> 3028                         raise ValueError('Lengths must match to compare')
       3029                 func = getattr(self.values, op)
       3030                 result = func(np.asarray(other))

    ValueError: Lengths must match to compare

#### Expected Output

My preference would be

    In [6]: a.dtype
    Out[6]: dtype('int64')

... but if one prioritises coherence with ``a = a + b``, then the expected behaviour would probably be

    In [6]: a.dtype
    Out[6]: dtype('float64')
   
    In [7]: a.index == b.index
    Out[7]: array([ True,  True,  True,  True], dtype=bool)

(outer join). Instead I think the current behaviour is hard to rationalise (i.e. it is dictated by implementation).

#### output of ``pd.show_versions()``

    In [7]: pd.show_versions()

    INSTALLED VERSIONS
    ------------------
    commit: 22cf50bbef7f28c63941328ea47502f1b202bda2
    python: 2.7.10.final.0
    python-bits: 64
    OS: Linux
    OS-release: 4.4.0-1-amd64
    machine: x86_64
    processor: 
    byteorder: little
    LC_ALL: None
    LANG: it_IT.utf8

    pandas: 0.18.0+48.g22cf50b.dirty
    nose: 1.3.6
    pip: 1.5.6
    setuptools: 18.4
    Cython: 0.23.2
    numpy: 1.10.4
    scipy: 0.16.0
    statsmodels: 0.8.0.dev0+755fa81
    xarray: None
    IPython: 2.4.1
    sphinx: 1.3.1
    patsy: 0.3.0-dev
    dateutil: 2.2
    pytz: 2012c
    blosc: None
    bottleneck: None
    tables: 3.2.2
    numexpr: 2.4.3
    matplotlib: 1.5.0rc2
    openpyxl: 2.3.0
    xlrd: 0.9.4
    xlwt: 0.7.5
    xlsxwriter: 0.7.3
    lxml: 3.4.4
    bs4: 4.4.0
    html5lib: 0.999
    httplib2: 0.9.1
    apiclient: None
    sqlalchemy: 1.0.11
    pymysql: None
    psycopg2: 2.6.1 (dt dec mx pq3 ext lo64)
    jinja2: 2.8
    boto: 2.38.0
    pandas_datareader: None

**Comment by jreback** [comment]
operating on series involves ``.align`` them, basically they are reindexed to the union.

```
In [23]: a = pd.Series(1, index=range(3))

In [24]: a = pd.Series(1, index=range(3))

In [25]: a2, b2 = a.align(b)

In [26]: a2
Out[26]: 
0    1.0
1    1.0
2    1.0
3    NaN
dtype: float64

In [27]: b2
Out[27]: 
0    2
1    2
2    2
3    2
dtype: int64

So ``a+b`` is exactly right
```
In [18]: b = pd.Series(2, index=range(4))

In [19]: a = pd.Series(1, index=range(3))

In [20]: a+b
Out[20]: 
0    3.0
1    3.0
2    3.0
3    NaN
dtype: float64
```

Looks like this is just taking the a indices (and not aligning). This probably doesn't have as much test coverage. So I would call this a bug in that ``a`` should yield ``a+b`` here.
```
In [21]: a+=b

In [22] a`
Out[22]: 
0    3.0
1    3.0
2    3.0
dtype: float64
```

**Comment by toobaz** [comment]
True, your proposed solution is coherent with ``a = a+b``. Just one more argument in favour of my preferred one: coherence with ``df['a'] = b``:

    In [2]: df = pd.DataFrame(index=range(3))

    In [3]: df['a'] = 1

    In [4]: b = pd.Series(5, index=range(4))

    In [5]: df['a'] = b

    In [6]: df
    Out[6]: 
       a
    0  5
    1  5
    2  5

... then, according to your solution (trivially extended to DataFrames), we would have:

    In [7]: df['a'] += b

    In [8]: df
    Out[8]: 
          a
    0  10.0
    1  10.0
    2  10.0
    3  NaN

... while according to my proposal, we would have:

    In [7]: df['a'] += b

    In [8]: df
    Out[8]: 
          a
    0  10
    1  10
    2  10

Notice that with neither of these ``df['a'] += b`` has the same effect of ``df['a'] = a + b`` anyway - which is instead true with the currently implemented behaviour. In other words: whatever decision is taken, there is some inconsistency. I like the option of not changing the dtype just because I think it is more useful.

**Comment by jreback** [comment]
hmm, there might be some logic in an in-place op to always do left-align (only). I don't think we want to make
``df['a'] += b`` change the index as the ``b`` should get aligned to the ``df.index``, NOT the union.

so on second thought I don't think this is a bug, but rather an explict design decision.

**Comment by toobaz** [comment]
I agree with you. So coming back to the dtype, OK for not changing it (when the casting is only a consequence of missing values)? Or talking about implementation: OK for calculating the operation **on the left axis**, *and then* putting the values in the left series (rather than calculating the operation on the union and then putting only values corresponding to the left axis in the left series)?

**Comment by jreback** [comment]
@toobaz I dont really recall this code path. If you can do this w/o jumping thru hoops then ok with changing impl to NOT change dtype. might be straghtforward.
